### PR TITLE
[CB] [Major] Add tensor paralellism

### PR DESCRIPTION
**Pull request by remi-or**
This PR adds support for TP in continuous batching. The major changes required to do this were:
- add inter-process communication for the requests states
- add per TP group seeding
- add hints to prevent NCCL graph mixing
- change hash function to avoid python `hash` which is salted depending on the process

It also adds a mechanism to the benchmark script to make sure the generation is coherent. 

## Performance

| Benchmark | TP1 main tok/s | TP1 tok/s | TP2 tok/s | Speedup | TP1 acc | TP2 acc |
|---|---|---|---|---|---|---|
| gsm8k_default | 2,454 | 2,454 | 3,657 | 1.49× | 0.822 | 0.819 |
| gsm8k_sampling | 1,940 | 1,942 | 2,616 | 1.35× | 0.792 | 0.775 |
| gsm8k_compile | 2,463 | 2,467 | 3,689 | 1.50× | 0.822 | 0.821 |
| gsm8k_no_fast_decode | 2,367 | 2,370 | 3,522 | 1.49× | 0.822 | 0.819 |
| gsm8k_bare_bones | 1,877 | 1,881 | 2,331 | 1.24× | 0.821 | 0.821 |
| ifeval_default | 7,890 | 7,898 | 15,135 | 1.92× | 0.442 | 0.455 |
| rollouts_1024 | 3,200 | 3,199 | 4,281 | 1.34× | — | — |
| rollouts_2048 | 3,048 | 3,049 | 4,194 | 1.38× | — | — |
| rollouts_4096 | 2,719 | 2,719 | 3,887 | 1.43× | — | — |
| rollouts_8192 | 2,209 | 2,211 | 3,345 | 1.51× | — | — |
| rollouts_16384 | 1,465 | 1,465 | 2,589 | 1.77× | — | — |
| few_blocks | 686 | 696 | 840 | 1.21× | — | — |
| multi_return_seq | 1,544 | 1,553 | 1,926 | 1.24× | — | — |

No perf regression, TP is faster.

## Tests

Added tests for TP, all tests run.

### CODE REVIEW

File: src/transformers/generation/continuous_batching/cache.py [hunk 1]
```diff
@@ -163,14 +163,14 @@ def __init__(
                 self.layer_index_to_group_indices[layer] = (i, j)
                 self.sliding_windows[layer] = sliding_window
 
-        # Handle TP (or dont)
+        # Account for TP: each KV head is dispatched to a different GPU, so the effective number of KV heads per GPU is
+        # simply divided by the TP size (number of GPUs)
         if tp_size is not None and tp_size > 1:
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/cache.py]: only if the attention k and v are target of the tp plan tho

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 2]
```diff
@@ -156,28 +159,35 @@ def __init__(
             generation_config: The generation configuration
             continuous_batching_config: The continuous batching configuration
             logit_processor: The [`ContinuousBatchingLogitsProcessorList`] object used to process the logits.
-            input_queue: Queue for incoming requests
+            input_queue: Queue for incoming requests. Is None if this process is not a TP driver.
+            cancel_queue: Queue for cancellation request_ids. Is None if this process is not a TP driver.
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/continuous_api.py:163]: Okay will read the rest to see if all are drivers or not

File: src/transformers/generation/continuous_batching/continuous_api.py [hunk 3]
```diff
@@ -248,22 +258,32 @@ def reset(self) -> None:
 
     @traced
     def _get_new_requests(self) -> None:
-        """Pull new requests from the input queue and add to waiting list."""
-        while not self.input_queue.empty():
+        """Pull new requests and cancellations from the queues and apply them to the scheduler. If the process is a TP
+        driver, the input_queue and cancel_queue are not None and the process will drain them. Otherwise, the process
+        will wait for the TP driver to send a payload containing the new requests and cancellations."""
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/continuous_api.py]: okay

File: src/transformers/generation/continuous_batching/utils.py [hunk 4]
```diff
@@ -204,3 +208,103 @@ def create_warmup_future_states(
...
+            self.ingress_group = dist.new_group(ranks=dist.get_process_group_ranks(self.tp_group), backend="gloo")
+        else:
+            self.tp_size = 1
+            self.tp_group = None
+            self.tp_root_global_rank = 0
+            self.tp_local_rank = 0
+            self.ingress_group = None
+
+        # The TP-driver is the rank that owns the request queue and the scheduler decisions inside its TP group.
+        self.is_tp_driver = self.tp_local_rank == 0
```
Comment (ArthurZucker) [inline src/transformers/generation/continuous_batching/utils.py]: Let's make a func for easy overwrite? IDK if there can be an advantage but let's give freedom?

**Review by ArthurZucker — APPROVED** [review]
LGTM be careful, EP only will fail if you devide by default with tp size the head dim

File: src/transformers/generation/continuous_batching/cache.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cache.py]: What could be the other targets? Not familiar enough with the TP plan tbh

File: src/transformers/generation/continuous_batching/continuous_api.py (hunk 2 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/continuous_api.py:163]: ?

File: src/transformers/generation/continuous_batching/utils.py (hunk 4 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/utils.py]: Why not, adding it.

File: src/transformers/generation/continuous_batching/cache.py (hunk 1 above)
Comment (remi-or) [reply src/transformers/generation/continuous_batching/cache.py]: Ok, added a boolean `kv_is_tp = "layers.*.self_attn.k_proj" in config.tp_plan and "layers.*.self_attn.v_proj" in config.tp_plan` to condition this.
